[Pipeline] Fix nil pointer when accessing config

eddycharly opened the pull request:
This PR fixes https://github.com/kubernetes/test-infra/issues/16120

When the prow config is not set on the command line, the config returned by the config agent is `nil`.
This must be verified before accessing properties and use fallback/default values when it is `nil`.

eddycharly commented:
/assign @cblecker

matthyx commented:
/ok-to-test

matthyx commented:
I am not sure I like the current implementation... isn't there a better way to deal with default configs?

eddycharly commented:
One option would be to make the command line argument mandatory.
Having the agent manage a default config would be great but it would require more changes.

What do you prefer ?

matthyx commented:
I would make it mandatory.
/cc @cjwagner

stevekuznetsov commented:
Is it valid to run Prow without a config?

eddycharly commented:
I just found the config.ConfigPath function, I think it should do the trick.

eddycharly commented:
In the end i modified `pipeline/dev.yaml` to mount the prow config and added a call to `config.ConfigPath` to support default config path.

I reverted the changes where i check for `nil` when getting the config.

matthyx commented:
Looks 100x better, thanks!
/lgtm

cblecker commented:
This looks okay to me, but I'd like to get some more eyes on it.

/cc @stevekuznetsov @cjwagner @fejta 
/hold

stevekuznetsov commented on prow/cmd/pipeline/main.go:
```diff
@@ -64,18 +64,19 @@ func (o *options) parse(flags *flag.FlagSet, args []string) error {
 	flags.StringVar(&o.totURL, "tot-url", "", "Tot URL")
 	flags.StringVar(&o.kubeconfig, "kubeconfig", "", "Path to kubeconfig. Only required if out of cluster")
-	flags.StringVar(&o.config, "config", "", "Path to prow config.yaml")
+	flags.StringVar(&o.configPath, "config", "", "Path to prow config.yaml")
```
We should add a validation that this is set

eddycharly replied:
I didn’t add validation here because I call config.ConfigPath later that sets the default config path if it was not set on the command line.

It is what is done in other components (hook, deck, etc...) and I thought it would be more consistent to do the same thing here. 

I can add validation if you prefer though.

cjwagner replied:
It is bad that other components default this field. We only continue to do that for backwards compatibility and it looks like that should have been deprecated by now so please don't add more components that use it: https://github.com/kubernetes/test-infra/blob/7ac63bdbab0f3666497bede973d7703bb30e9cd4/prow/config/config.go#L1324
Validation is preferable.

matthyx replied:
Sorry, didn't know that... I should probably follow more closely sig meetings where we communicate this -_-"

eddycharly replied:
Thank you for your explanation, i removed the call to `config.ConfigPath` and added validation.

fejta commented on prow/cmd/pipeline/main_test.go:
```diff
@@ -29,18 +29,25 @@ func TestOptions(t *testing.T) {
 		err      bool
 	}{{
-		name:     "defaults work",
+		name:     "defaults doesn't work (--config argument is mandatory)",
```
Please fix this so it provides the minimal amount necessary to work (aka add --config)

eddycharly replied:
I enhanced the validation message and the test name with a more explicit message about what should be done to make it work.
